File: python/DataStructures/LinkedList/LinkedList.py
from __future__ import annotations
from typing import Tuple, TypeVar, Generic, Callable

T = TypeVar('T')


class ListItem(Generic[T]):
    __value: T
    __next_item: ListItem[T] or None

    def __init__(self, value: T):
        self.__value = value
        self.__next_item = None

    def set_next(self, next_item: T):
        self.__next_item = next_item

    def get_value(self) -> T:
        return self.__value

    def get_next_item(self) -> ListItem[T]:
        return self.__next_item

    def __repr__(self):
        return str(self.__value)


class LinkedListIterator(Generic[T]):
    __current: ListItem[T]
    __index: int

    def __init__(self, start_item: ListItem[T]):
        self.__current = start_item
        self.__index = 0

    def __next__(self) -> Tuple[T, int]:
        if self.__current is None:
            raise StopIteration
        item: ListItem[T] = self.__current
        index: int = self.__index
        self.__current = item.get_next_item()
        self.__index += 1
        return item.get_value(), index


class LinkedList(Generic[T]):
    __start_item: ListItem[T] or None

    def __init__(self):
        self.__start_item = None

    def __iter__(self) -> T:
        return LinkedListIterator(self.__start_item)

    def is_empty(self) -> bool:
        return self.__start_item is None

    def insert(self, index: int, item: T) -> bool:
        new_item: ListItem[T] = ListItem(item)
        if index == 0:
            new_item.set_next(self.__start_item)
            self.__start_item = new_item
            return True
        else:
            _index: int = 1
            curr_item: ListItem[T] = self.__start_item
            while curr_item is not None:
                if _index == index:
                    new_item.set_next(curr_item.get_next_item())
                    curr_item.set_next(new_item)
                    return True
                curr_item = curr_item.get_next_item()
                _index += 1
        return False

    def append(self, item: T):
        if self.__start_item is None:
            self.__start_item = ListItem(item)
        else:
            curr_item: ListItem[T] = self.__start_item
            while curr_item.get_next_item() is not None:
                curr_item = curr_item.get_next_item()
            curr_item.set_next(ListItem[T](item))

    def get(self, index: int) -> T or None:
        for item, _index in self:
            if index == _index:
                return item
        return None

    def remove(self, index: int) -> bool:
        if index == 0:
            if self.__start_item is not None:
                self.__start_item = self.__start_item.get_next_item()
                return True
        else:
            _index: int = 1
            curr_item: ListItem[T] = self.__start_item
            while curr_item.get_next_item() is not None:
                if _index == index:
                    _to_remove: ListItem[T] = curr_item.get_next_item()
                    if _to_remove is not None:
                        curr_item.set_next(_to_remove.get_next_item())
                        return True
                    else:
                        return False
                curr_item = curr_item.get_next_item()
                _index += 1
        return False

    def iterate_items(self, callback: Callable[[T, int], None]):
        curr_item: ListItem[T] = self.__start_item
        index: int = 0
        while curr_item is not None:
            callback(curr_item.get_value(), index)
            curr_item = curr_item.get_next_item()

    def __repr__(self):
        as_str: str = ''
        curr_item: ListItem[T] = self.__start_item
        while curr_item is not None:
            as_str += ' ' + str(curr_item.get_value())
            curr_item = curr_item.get_next_item()
        return as_str

File: lib/src/dataStructures/linkedList/LinkedList.ts
import LinkedItem from "./LinkedItem";
import { MatchFunction, Optional } from "../../types";
import { ObservableVersioned } from "../../common";
import { NO_MATCH } from "../../algorithms/search/common";

export default class LinkedList<T> extends ObservableVersioned {
  startItem: Optional<LinkedItem<T>>;
  length: number;

  constructor() {
    super();
    this.startItem = undefined;
    this.length = 0;
  }

  *[Symbol.iterator]() {
    let cItem = this.startItem;
    while (!!cItem) {
      yield cItem.getValue();
      cItem = cItem.getNextItem();
    }
  }

  size(): number {
    return this.length;
  }

  toArray() {
    const arr: T[] = [];
    for (const i of this) {
      arr.push(i);
    }
    return arr;
  }

  isEmpty() {
    return this.length === 0;
  }

  empty(): this {
    this.startItem = undefined;
    this.length = 0;
    return this;
  }

  insert(index: number, item: T) {
    let inserted = false;
    const newItem: LinkedItem<T> = new LinkedItem<T>(item);

    if (index === 0) {
      newItem.setNextItem(this.startItem);
      this.startItem = newItem;
      inserted = true;
    } else {
      let tIndex: number = 1;
      let currentItem: LinkedItem<T> = this.startItem;
      while (!!currentItem) {
        if (tIndex === index) {
          newItem.setNextItem(currentItem.getNextItem());
          currentItem.setNextItem(newItem);
          inserted = true;
          break;
        }

        tIndex += 1;
        currentItem = currentItem.getNextItem();
      }
    }

    if (inserted) {
      this.length += 1;
      this.tickVersion();
    }

    return inserted;
  }

  get(index: number): T {
    let tIndex = 0;
    let cItem = this.startItem;
    while (!!cItem) {
      if (tIndex === index) {
        return cItem.getValue();
      }
      cItem = cItem.getNextItem();
      tIndex += 1;
    }

    return undefined;
  }

  appendAll(...items: T[]): this {
    items.forEach(i => this.append(i));
    return this;
  }

  append(item: T): this {
    const newItem = new LinkedItem(item);

    if (!!this.startItem) {
      let currentItem = this.startItem;

      while (!!currentItem.getNextItem()) {
        currentItem = currentItem.getNextItem();
      }

      currentItem.setNextItem(newItem);
    } else {
      this.startItem = newItem;
    }

    this.length += 1;
    this.tickVersion();
    return this;
  }

  contains(matchFunction: MatchFunction<T>): boolean {
    return this.findMatch(matchFunction) !== NO_MATCH;
  }

  findMatch(matchFunction: MatchFunction<T>): number {
    let index: number = 0;
    let cItem: LinkedItem<T> = this.startItem;
    while (!!cItem) {
      if (matchFunction(cItem.getValue())) {
        return index;
      }
      cItem = cItem.getNextItem();
      index += 1;
    }
    this.tickVersion();

    return NO_MATCH;
  }

  /**
   * Removes
   * @param {function} matchFunction Returns true for the object being removed
   * @returns The removed item
   */
  removeMatch(matchFunction: MatchFunction<T>): Optional<T> {
    let removed: Optional<T>;
    let index: number = 0;
    let cItem: LinkedItem<T> = this.startItem;
    while (!!cItem) {
      if (matchFunction(cItem.getValue())) {
        removed = this.remove(index);
        break;
      }
      cItem = cItem.getNextItem();
      index += 1;
    }
    this.tickVersion();

    return removed;
  }

  /**
   * Remove an item in a specific position
   * @param {number} index The index of the object to remove
   * @returns The removed item
   */
  remove(index: number): T {
    let removed: Optional<T>;

    if (index === 0) {
      if (!!this.startItem) {
        removed = this.startItem.getValue();
        this.startItem = this.startItem.getNextItem();
      }
    } else {
      let tIndex = 1;
      let currentItem = this.startItem;
      while (!!currentItem.getNextItem()) {
        if (tIndex === index) {
          const toRemove = currentItem.getNextItem();
          if (!!toRemove) {
            removed = toRemove.getValue();
            currentItem.setNextItem(toRemove.getNextItem());
            break;
          }
        }
        currentItem = currentItem.getNextItem();
        tIndex += 1;
      }
    }

    if (removed) {
      this.length -= 1;
      this.tickVersion();
    }

    return removed;
  }

  toString(): string {
    // return [this].reduce((acc, curr) => (acc += ` ${curr.getValue()}`), "");
    const arr: T[] = [];
    for (const i of this) {
      arr.push(i);
    }
    return arr.join(" ");
  }
}
